README.md
# ImageProcessing
#### Experiments with processing images to make cooler, more exciting images. ;)


personal_sobel_experiment/CHANGELOG
1/17/2019 - Downloaded files and edited the code's style to match my own.
1/20/2019 - Finished the program.
1/21/2019 - Wrote a Makefile, this CHANGELOG, and the README.
1/21/2019 - Experimented with functions, did not keep.
1/22/2019 - Verified working code with bigger_digger and motorcycle.
1/22/2019 - Experimented on non-.pgm images as well as non-1024*1024 images (poor results).
1/23/2019 - Added user input and fixed some errors with the help of the TA, Le Chan.

ee285_experiments/Img.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <math.h>
#include "Img.h"

// Global Definitions
#define IMG_ROWS 900
#define IMG_COLS 900

// Global Variables
uint16_t bpp;
uint32_t pixel_arr_loc;
int row, col;

// Pixel type definition
typedef struct pixel {
	unsigned int r, g, b, a;
} pixel_type;

pixel_type in[IMG_ROWS][IMG_COLS];
pixel_type out[IMG_ROWS][IMG_COLS];

int load_image(FILE *fp) {
	unsigned char signature[2];
	uint32_t file_size;
	uint32_t dib_size;
	uint16_t comp_method;

	if(fp == NULL) {
		printf("Unable to open file.\n");
		return 1;
	}

	fread(signature, 1, 2, fp);
	printf("sig %c %c\n", signature[0], signature[1]);
	if(signature[0] != 'B' || signature[1] != 'M') {
		printf("Not a valid BMP file.\n");
		return 1;
	}

	fseek(fp, 2, SEEK_SET);
	fread(&file_size, 4, 1, fp);
	printf("file size %u\n", file_size);

	fseek(fp, 10, SEEK_SET);
	fread(&pixel_arr_loc, 4, 1, fp);
	printf("pixel array location %u\n", pixel_arr_loc);

	fseek(fp, 14, SEEK_SET);
	fread(&dib_size, 4, 1, fp);
	printf("dib size %u\n", dib_size);
	if(dib_size != 40) {
		printf("BMP version not supported.\n");
		return 1;
	}

	fseek(fp, 28, SEEK_SET);
	fread(&bpp, 2, 1, fp);
	printf("bits per pixel %u\n", bpp);
	if(bpp != 32) {
		printf("Bits per pixel not supported.\n");
		return 1;
	}

	fseek(fp, 30, SEEK_SET);
	fread(&comp_method, 2, 1, fp);
	printf("compression method %u\n", comp_method);
	if(comp_method != 0) {
		printf("Compression method not supported.\n");
		return 1;
	}

	int bytes_per_row = IMG_COLS * bpp / 8;
	int padding = 0;
	if(bytes_per_row % 4 != 0) {
		padding = 4 - bytes_per_row % 4;
	}
	printf("input padding %u\n", padding);
	fseek(fp, pixel_arr_loc, SEEK_SET);
	int row;
	int col;
	for(row = IMG_ROWS - 1; row >= 0; row--) {
		for(col = 0; col < IMG_COLS; col++) {
			fread(&in[row][col].b, 1, 1, fp);
			fread(&in[row][col].g, 1, 1, fp);
			fread(&in[row][col].r, 1, 1, fp);
			fread(&in[row][col].a, 1, 1, fp);
			out[row][col].b = in[row][col].b;
			out[row][col].g = in[row][col].g;
			out[row][col].r = in[row][col].r;
			out[row][col].a = in[row][col].a;
			fseek(fp, padding, SEEK_CUR);
		}
	}
	return 0;
}

int write_image(FILE *outf, FILE *in) {
	int c;

	rewind(outf);
	rewind(in);

	c = fgetc(in);
	while(c != EOF) {
		fputc(c, outf);
		c = fgetc(in);
	}

	int bytes_per_row = IMG_COLS * bpp / 8;
	int padding = 0;
	if(bytes_per_row % 4 != 0) {
		padding = 4 - bytes_per_row % 4;
	}
	int row;
	int col;
	printf("output padding %u\n", padding);
	fseek(outf, pixel_arr_loc, SEEK_SET);
	for(row = IMG_ROWS - 1; row >= 0; row--) {
		for( col = 0; col < IMG_COLS; col++) {
			fwrite(&out[row][col].b, 1, 1, outf);
			fwrite(&out[row][col].g, 1, 1, outf);
			fwrite(&out[row][col].r, 1, 1, outf);
			fwrite(&out[row][col].a, 1, 1, outf);
			fseek(outf, padding, SEEK_CUR);
		}
	}
	return 0;
}

void filter_grayscale() {

	for(row=0; row<IMG_ROWS; row++) {
		for(col=0; col<IMG_COLS; col++) {
			if(in[row][col].r > in[row][col].g && in[row][col].r > in[row][col].b) {
			//	out[row][col].r = in[row][col].r;
				out[row][col].g = in[row][col].r;
				out[row][col].b = in[row][col].r;
			} else if(in[row][col].g > in[row][col].r && in[row][col].g > in[row][col].b) {
				out[row][col].r = in[row][col].g;
			//	out[row][col].g = in[row][col].g;
				out[row][col].b = in[row][col].g;
			} else if(in[row][col].b > in[row][col].r && in[row][col].b > in[row][col].g) {
				out[row][col].r = in[row][col].g;
				out[row][col].g = in[row][col].g;
			//	out[row][col].b = in[row][col].g;
			}
		}
	}
	
}

void filter_red_balloon() {
	
	for(row=0; row<IMG_ROWS; row++) {
		for(col=0; col<IMG_COLS; col++) {
			if(close_to_red(in[row][col].r, in[row][col].g, in[row][col].b) == 0) {
				if(in[row][col].r > in[row][col].g && in[row][col].r > in[row][col].b) {
					out[row][col].g = in[row][col].r;
					out[row][col].b = in[row][col].r;
				} else if(in[row][col].g > in[row][col].r && in[row][col].g > in[row][col].b) {
					out[row][col].r = in[row][col].g;
					out[row][col].b = in[row][col].g;
				} else if(in[row][col].b > in[row][col].r && in[row][col].b > in[row][col].g) {
					out[row][col].r = in[row][col].g;
					out[row][col].g = in[row][col].g;
				}
			} else {
				out[row][col].r = in[row][col].r;
				out[row][col].g = in[row][col].g;
				out[row][col].b = in[row][col].b;
			}
		}
	}
	
}

int close_to_red(unsigned int red, unsigned int green, unsigned int blue) {
	const double R = 255.0;
	const double G = 0.0;
	const double B = 0.0;
	const double R_tolerance = 125.0;
	const double G_B_tolerance = 50.0;
	
	if(fabs((double)red-255.0) < R_tolerance && fabs((double)green-0.0) < G_B_tolerance && fabs((double)blue-0.0) < G_B_tolerance) {
		return 1;
	} else {
		return 0;
	}
}

ee285_experiments/lab8.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <math.h>

// Global Definitions
#define IMG_ROWS 900
#define IMG_COLS 900

// Global Variables
char input_file_name[] = "cy.bmp";
char output_file_name[] = "cy2.bmp";
uint16_t bpp;
uint32_t pixel_arr_loc;
int row, col;

typedef struct pixel {
	unsigned int r, g, b, a;
} pixel_type;

pixel_type in[IMG_ROWS][IMG_COLS];
pixel_type out[IMG_ROWS][IMG_COLS];

// Functions declarations
int load_image(FILE *fp);
int write_image(FILE *outf, FILE *in);
void filter_grayscale();
void filter_red_balloon();
int close_to_red(unsigned int red, unsigned int green, unsigned int blue);

// Main Function
int main() {

	FILE *in = fopen(input_file_name, "rb");
	FILE *out = fopen(output_file_name, "wb");
	load_image(in);

//	filter_red_balloon();
	filter_grayscale();
	
	write_image(out, in);
	fclose(in);
	fclose(out);
	
	return 0;
}

int load_image(FILE *fp) {
	unsigned char signature[2];
	uint32_t file_size;
	uint32_t dib_size;
	uint16_t comp_method;

	if(fp == NULL) {
		printf("Unable to open file.\n");
		return 1;
	}

	fread(signature, 1, 2, fp);
	printf("sig %c %c\n", signature[0], signature[1]);
	if(signature[0] != 'B' || signature[1] != 'M') {
		printf("Not a valid BMP file.\n");
		return 1;
	}

	fseek(fp, 2, SEEK_SET);
	fread(&file_size, 4, 1, fp);
	printf("file size %u\n", file_size);

	fseek(fp, 10, SEEK_SET);
	fread(&pixel_arr_loc, 4, 1, fp);
	printf("pixel array location %u\n", pixel_arr_loc);

	fseek(fp, 14, SEEK_SET);
	fread(&dib_size, 4, 1, fp);
	printf("dib size %u\n", dib_size);
	if(dib_size != 40) {
		printf("BMP version not supported.\n");
		return 1;
	}

	fseek(fp, 28, SEEK_SET);
	fread(&bpp, 2, 1, fp);
	printf("bits per pixel %u\n", bpp);
	if(bpp != 32) {
		printf("Bits per pixel not supported.\n");
		return 1;
	}

	fseek(fp, 30, SEEK_SET);
	fread(&comp_method, 2, 1, fp);
	printf("compression method %u\n", comp_method);
	if(comp_method != 0) {
		printf("Compression method not supported.\n");
		return 1;
	}

	int bytes_per_row = IMG_COLS * bpp / 8;
	int padding = 0;
	if(bytes_per_row % 4 != 0) {
		padding = 4 - bytes_per_row % 4;
	}
	printf("input padding %u\n", padding);
	fseek(fp, pixel_arr_loc, SEEK_SET);
	int row;
	int col;
	for(row = IMG_ROWS - 1; row >= 0; row--) {
		for(col = 0; col < IMG_COLS; col++) {
			fread(&in[row][col].b, 1, 1, fp);
			fread(&in[row][col].g, 1, 1, fp);
			fread(&in[row][col].r, 1, 1, fp);
			fread(&in[row][col].a, 1, 1, fp);
			out[row][col].b = in[row][col].b;
			out[row][col].g = in[row][col].g;
			out[row][col].r = in[row][col].r;
			out[row][col].a = in[row][col].a;
			fseek(fp, padding, SEEK_CUR);
		}
	}
	return 0;
}

int write_image(FILE *outf, FILE *in) {
	int c;

	rewind(outf);
	rewind(in);

	c = fgetc(in);
	while(c != EOF) {
		fputc(c, outf);
		c = fgetc(in);
	}

	int bytes_per_row = IMG_COLS * bpp / 8;
	int padding = 0;
	if(bytes_per_row % 4 != 0) {
		padding = 4 - bytes_per_row % 4;
	}
	int row;
	int col;
	printf("output padding %u\n", padding);
	fseek(outf, pixel_arr_loc, SEEK_SET);
	for(row = IMG_ROWS - 1; row >= 0; row--) {
		for( col = 0; col < IMG_COLS; col++) {
			fwrite(&out[row][col].b, 1, 1, outf);
			fwrite(&out[row][col].g, 1, 1, outf);
			fwrite(&out[row][col].r, 1, 1, outf);
			fwrite(&out[row][col].a, 1, 1, outf);
			fseek(outf, padding, SEEK_CUR);
		}
	}
	return 0;
}

void filter_grayscale() {

	for(row=0; row<IMG_ROWS; row++) {
		for(col=0; col<IMG_COLS; col++) {
			if(in[row][col].r > in[row][col].g && in[row][col].r > in[row][col].b) {
			//	out[row][col].r = in[row][col].r;
				out[row][col].g = in[row][col].r;
				out[row][col].b = in[row][col].r;
			} else if(in[row][col].g > in[row][col].r && in[row][col].g > in[row][col].b) {
				out[row][col].r = in[row][col].g;
			//	out[row][col].g = in[row][col].g;
				out[row][col].b = in[row][col].g;
			} else if(in[row][col].b > in[row][col].r && in[row][col].b > in[row][col].g) {
				out[row][col].r = in[row][col].g;
				out[row][col].g = in[row][col].g;
			//	out[row][col].b = in[row][col].g;
			}
		}
	}
	
}

void filter_red_balloon() {
	
	for(row=0; row<IMG_ROWS; row++) {
		for(col=0; col<IMG_COLS; col++) {
			if(close_to_red(in[row][col].r, in[row][col].g, in[row][col].b) == 0) {
				if(in[row][col].r > in[row][col].g && in[row][col].r > in[row][col].b) {
					out[row][col].g = in[row][col].r;
					out[row][col].b = in[row][col].r;
				} else if(in[row][col].g > in[row][col].r && in[row][col].g > in[row][col].b) {
					out[row][col].r = in[row][col].g;
					out[row][col].b = in[row][col].g;
				} else if(in[row][col].b > in[row][col].r && in[row][col].b > in[row][col].g) {
					out[row][col].r = in[row][col].g;
					out[row][col].g = in[row][col].g;
				}
			} else {
				out[row][col].r = in[row][col].r;
				out[row][col].g = in[row][col].g;
				out[row][col].b = in[row][col].b;
			}
		}
	}
	
}

int close_to_red(unsigned int red, unsigned int green, unsigned int blue) {
	const double R = 255.0;
	const double G = 0.0;
	const double B = 0.0;
	const double R_tolerance = 125.0;
	const double G_B_tolerance = 50.0;
	
	if(fabs((double)red-255.0) < R_tolerance && fabs((double)green-0.0) < G_B_tolerance && fabs((double)blue-0.0) < G_B_tolerance) {
		return 1;
	} else {
		return 0;
	}
}

personal_sobel_experiment/sobel.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <math.h>

// @author Steven Marshall Sheets - SMSheets@IAState.edu

// gcc -g -lm -o sobel.out sobel.c
// ./sobel.out

// https://www.youtube.com/watch?v=uihBwtPIBxM

/* Do not modify write_pgm() or read_pgm() */
int write_pgm(char *file, void *image, uint32_t x, uint32_t y) {
    FILE *o;

    if (!(o = fopen(file, "w"))) {
        perror(file);
        return -1;
    }

    fprintf(o, "P5\n%u %u\n255\n", x, y);

    /* Assume input data is correctly formatted. *
    *  There's no way to handle it, otherwise.   */

    if (fwrite(image, 1, x * y, o) != (x * y)) {
        perror("fwrite");
        fclose(o);
        return -1;
    }

    fclose(o);

    return 0;
}

/* A better implementation of this function would read the image dimensions *
 * from the input and allocate the storage, setting x and y so that the     *
 * user can determine the size of the file at runtime.  In order to         *
 * minimize complication, I've written this version to require the user to  *
 * know the size of the image in advance.                                   */
int read_pgm(char *file, void *image, uint32_t x, uint32_t y) {
    FILE *f;
    char s[80];
    unsigned i, j;

    if (!(f = fopen(file, "r"))) {
        perror(file);
        return -1;
    }

    if (!fgets(s, 80, f) || strncmp(s, "P5", 2)) {
        fprintf(stderr, "Expected P6\n");
        return -1;
    }

    /* Eat comments */
    do {
        fgets(s, 80, f);
    } while (s[0] == '#');

    if (sscanf(s, "%u %u", &i, &j) != 2 || i != x || j != y) {
        fprintf(stderr, "Expected x and y dimensions %u %u\n", x, y);
        fclose(f);
        return -1;
    }

    /* Eat comments */
    do {
        fgets(s, 80, f);
    } while (s[0] == '#');

    if (strncmp(s, "255", 3)) {
        fprintf(stderr, "Expected 255\n");
        fclose(f);
        return -1;
    }

    if (fread(image, 1, x * y, f) != x * y) {
        perror("fread");
        fclose(f);
        return -1;
    }

    fclose(f);

    return 0;
}

int main(int argc, char *argv[]) {
    int8_t image[1024][1024];
    int8_t out[1024][1024];

    /* Example usage of PGM functions */
    /* This assumes that motorcycle.pgm is a pgm image of size 1024 x 1024 */
    //read_pgm("motorcycle.pgm", image, 1024, 1024);

    char filename[50];    // Thank you Le Chen for the help!
    printf("Please enter the file name with the file extension: \n");
    scanf("%s", filename);
    read_pgm(filename, image, 1024, 1024);

    const int matrix_length = 3;
    const int x_transform[3][3] = {{-1, 0, 1},
                                   {-2, 0, 2},
                                   {-1, 0, 1}};
    const int y_transform[3][3] = {{-1, -2, -1},
                                   {0,  0,  0},
                                   {1,  2,  1}};

    int r, c, j, i;
    for (r = 0; r < 1024; r++) {
        for (c = 0; c < 1024; c++) {
            int x_accumulator = 0;    // For some odd reason, declaring these on the same
            int y_accumulator = 0;    // line produces drastically different results.

            for (j = 0; j < matrix_length; j++) {
                for (i = 0; i < matrix_length; i++) {
                    int row = r + (j - ceil(matrix_length / 2));
                    int col = c + (i - ceil(matrix_length / 2));

                    x_accumulator += x_transform[j][i] * image[row][col];
                    y_accumulator += y_transform[j][i] * image[row][col];
                }
            }

            out[r][c] = sqrt((x_accumulator * x_accumulator) + (y_accumulator * y_accumulator));
        }
    }

    /* After processing the image and storing your  *
     * output in "out", write to motorcycle.edge.pgm. 	*/
    int length = strlen(filename);
    int x = 0;
    char file2[50];    // Thanks Le Chen!
    while (x < strlen(filename) - 4) {
        file2[x] = filename[x];
        x++;
    }

    write_pgm(strcat(file2, ".edge.pgm"), out, 1024, 1024);

    return 0;
}

cs327_sobel_filter/jeremy_solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <math.h>

/* Do not modify write_pgm() or read_pgm() */
int write_pgm(char *file, void *image, uint32_t x, uint32_t y) {
    FILE *o;

    if (!(o = fopen(file, "w"))) {
        perror(file);

        return -1;
    }

    fprintf(o, "P5\n%u %u\n255\n", x, y);

    /* Assume input data is correctly formatted. *
     * There's no way to handle it, otherwise.   */

    if (fwrite(image, 1, x * y, o) != (x * y)) {
        perror("fwrite");
        fclose(o);

        return -1;
    }

    fclose(o);

    return 0;
}

/* A better implementation of this function would read the image dimensions *
 * from the input and allocate the storage, setting x and y so that the     *
 * user can determine the size of the file at runtime.  In order to         *
 * minimize complication, I've written this version to require the user to  *
 * know the size of the image in advance.                                   */
int read_pgm(char *file, void *image, uint32_t x, uint32_t y) {
    FILE *f;
    char s[80];
    unsigned i, j;

    if (!(f = fopen(file, "r"))) {
        perror(file);

        return -1;
    }

    if (!fgets(s, 80, f) || strncmp(s, "P5", 2)) {
        fprintf(stderr, "Expected P5\n");

        return -1;
    }

    /* Eat comments */
    do {
        fgets(s, 80, f);
    } while (s[0] == '#');

    if (sscanf(s, "%u %u", &i, &j) != 2 || i != x || j != y) {
        fprintf(stderr, "Expected x and y dimensions %u %u\n", x, y);
        fclose(f);

        return -1;
    }

    /* Eat comments */
    do {
        fgets(s, 80, f);
    } while (s[0] == '#');

    if (strncmp(s, "255", 3)) {
        fprintf(stderr, "Expected 255\n");
        fclose(f);

        return -1;
    }

    if (fread(image, 1, x * y, f) != x * y) {
        perror("fread");
        fclose(f);

        return -1;
    }

    fclose(f);

    return 0;
}

int main(int argc, char *argv[]) {
    uint8_t image[1024][1024];
    int16_t x_filter[3][3] = {{-1, 0, 1},
                              {-2, 0, 2},
                              {-1, 0, 1}};
    int16_t y_filter[3][3] = {{-1, -2, -1},
                              {0,  0,  0},
                              {1,  2,  1}};
    uint8_t out[1024][1024];
    int16_t x_out, y_out, xy_out;
    int x, y;
    char *s;

    if (argc != 2) {
        fprintf(stderr, "Usage: %s <PGM file>\n", argv[0]);

        return -1;
    }

    read_pgm(argv[1], image, 1024, 1024);
    for (y = 0; y < 1024; y++) {
        for (x = 0; x < 1024; x++) {
            x_out = y_out = 0;
            if (x != 0 && y != 0 && x != 1023 && y != 1023) {
                x_out = ((((int16_t) image[y - 1][x - 1]) * x_filter[0][0]) +
                         (((int16_t) image[y - 1][x]) * x_filter[0][1]) +
                         (((int16_t) image[y - 1][x + 1]) * x_filter[0][2]) +
                         (((int16_t) image[y][x - 1]) * x_filter[1][0]) +
                         (((int16_t) image[y][x]) * x_filter[1][1]) +
                         (((int16_t) image[y][x + 1]) * x_filter[1][2]) +
                         (((int16_t) image[y + 1][x - 1]) * x_filter[2][0]) +
                         (((int16_t) image[y + 1][x]) * x_filter[2][1]) +
                         (((int16_t) image[y + 1][x + 1]) * x_filter[2][2]));
                y_out = ((((int16_t) image[y - 1][x - 1]) * y_filter[0][0]) +
                         (((int16_t) image[y - 1][x]) * y_filter[0][1]) +
                         (((int16_t) image[y - 1][x + 1]) * y_filter[0][2]) +
                         (((int16_t) image[y][x - 1]) * y_filter[1][0]) +
                         (((int16_t) image[y][x]) * y_filter[1][1]) +
                         (((int16_t) image[y][x + 1]) * y_filter[1][2]) +
                         (((int16_t) image[y + 1][x - 1]) * y_filter[2][0]) +
                         (((int16_t) image[y + 1][x]) * y_filter[2][1]) +
                         (((int16_t) image[y + 1][x + 1]) * y_filter[2][2]));
            }
            xy_out = round(sqrt((x_out * x_out) + (y_out * y_out)));
            if (xy_out > 255) {
                xy_out = 255;
            }
            out[y][x] = xy_out;
        }
    }

    /* Construct a string from the input image name with ".edge" stuck between *
     * the name and the extension.  5 bytes for ".edge" and the NULL byte.     */
    s = malloc(strlen(argv[0] + 6));
    strcpy(s, argv[1]);
    /* BUG: Fix it for a little challenge, but there's no bounty for this.  If  *
     * the input file name doesn't end with .pgm, or if it contains extra dots, *
     * this will not construct a correct output file name.  Correct solution is *
     * to iterate backwards through the file name and ensure that you've found  *
     * the last dot and that the file has the appropriate extension (which      *
     * could require a larger allocation).                                      */
    *strchr(s, '.') = '\0';
    strcat(s, ".edge.pgm");
    write_pgm(s, out, 1024, 1024);
    free(s);

    return 0;
}

/*
	I construct an output file name from the input file name.  This construction is slightly buggy, but you can fix it if you want (no bug bounty, though).
    See the comment in the source code for details. I don't use a loop for the convolution.  Instead, I implement the whole convolution in a single expression.
    The reason is entirely one of performance.  That the expanded, hard-coded expression is much, much, much faster than the little 3x3 loop that you probably wrote.
    Go ahead and time them to confirm.  You can time an application running in the shell in Lunux by running it using the "time" utility, e.g., "time sobel
    motorcycle.pgm".  This timer is not good enough for real scientific analysis, but it's plenty good enough for something like this. I cast all
    of my index operations into the input image to int16_t (a 16-bit, signed integer type) to avoid the sign and overflow issues that lead to "grainy" output.
*/